lib/deref.js

```diff
@@ -134,31 +140,50 @@ module.exports = {
       // not using _.get here because that fails if there's a . in the property name (Pagination.Envelope, for example)
       resolvedSchema = this._getEscaped(components, splitRef.slice(2));
       if (resolvedSchema) {
-        return this.resolveRefs(resolvedSchema, components, stack);
+        return this.resolveRefs(resolvedSchema, bodyTypeOption, components, stack);
       }
       return { value: 'reference ' + schema.$ref + ' not found in the api spec' };
     }
     if (schema.type === 'object' || schema.hasOwnProperty('properties')) {
       // go through all props
       schema.type = 'object';
       if (schema.hasOwnProperty('properties')) {
+        // shallow cloning schema object except properties object
+        let tempSchema = _.omit(schema, 'properties');
+        tempSchema.properties = {};
         for (prop in schema.properties) {
           if (schema.properties.hasOwnProperty(prop)) {
-            schema.properties[prop] = this.resolveRefs(schema.properties[prop], components, stack);
+            /* eslint-disable max-depth */
+            // handling OAS readOnly and writeOnly properties in schema
+            // Related Doc - https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#schemaObject
+            let property = schema.properties[prop];
+            if (property.readOnly && bodyTypeOption === BODY_TYPE.REQUEST) {
+              continue;
+            }
+            else if (property.writeOnly && bodyTypeOption === BODY_TYPE.RESPONSE) {
+              continue;
+            }
+            /* eslint-enable */
+            tempSchema.properties[prop] = property;
+            // eslint-disable-next-line max-len
```

Comment:
no need to disable `max-len` rule. Can move the extra part of the code down

Reply:
Can combine these two lines into:
`tempSchema.properties[prop] = this.resolveRefs(property, ...)`